Numerical cast

romainbrenguier opened the pull request:
This provides a unified way of converting mp_integer to numeric types, and could be extended to conversion from other types. To convert a numeric value `v` to a numeric type `t`, we would always use `numeric_cast<t>(v)`.
This will help implementing the changes suggested here:
https://github.com/diffblue/cbmc/pull/1572

romainbrenguier committed:
Add numeric_cast template for numeric conversion

This is for now only applied to mp_integer but could be extended to
other numerical types.

smowton reviewed (commented):
1. Don't add line breaks in the untouched part of the file
2. Use `static_assert` to check that the type you're casting to really is <= size of either `long long` or `unsigned long long`.
3. What about float and double? It looks like the signed case will allow them but their range is larger than int64, so this might behave strangely?

romainbrenguier commented:
@smowton 
For 3, I've added a check `is_integral` so that double and float are excluded.
For 2, added `static_assert` to check the max and min are within the bounds of the max and min of long long
For 1, removed the commit.

smowton commented:
One last thing -- use `decltype(mpi.to_long())` instead of `long long`, and similarly `mpi.to_ulong()`, thus keeping in sync even if we change `to_long()` to return `int128t` or whatever

LAJW reviewed (approved):
Looks alright.
